### PR TITLE
Add TimeConfig option to allow truncating of fractional seconds

### PR DESCRIPTION
**Pull request by adriangb**

### CODE REVIEW

**Comment by adriangb** [comment]
@samuelcolvin a couple of questions:
- Any input on how to fix the failing tests? It might take me a bit of work to actually grok what is going on during parsing, it's a lot of bytes, offsets, `length += i`, etc.
- Do you think we also need new `parse_str` methods? I was thinking not
- Should we call the parameter `time_config: TimeConfig` or add an empty `date_config: DateConfig` or are we certain we'll never want a config for dates?

File: src/time.rs [hunk 1]
```diff
@@ -488,3 +523,22 @@ impl PureTime {
         self.hour as u32 * 3_600 + self.minute as u32 * 60 + self.second as u32
     }
 }
+
+#[derive(Debug, Clone)]
+pub enum SecondsPrecisionOverflowBehavior {
+    Truncate,
+    Error,
+}
```
Comment (samuelcolvin) [inline src/time.rs:531]: might be good to implement `TryFrom<String>` for this here.

File: src/time.rs [hunk 2]
```diff
@@ -488,3 +523,22 @@ impl PureTime {
...
+
+#[derive(Debug, Clone)]
+pub struct TimeConfig {
+    pub seconds_precision_overflow_behavior: SecondsPrecisionOverflowBehavior,
+}
+
+impl Default for TimeConfig {
```
Comment (samuelcolvin) [inline src/time.rs:538]: better to do `derive(..., Default)` so you can do `SpeedateConfig {x: y, ...}`

File: src/datetime.rs [hunk 3]
```diff
@@ -270,6 +270,25 @@ impl DateTime {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use speedate::{DateTime, Date, Time, TimeConfig};
+    ///
+    /// let dt = DateTime::parse_bytes_with_config(b"2022-01-01T12:13:14Z", TimeConfig::default()).unwrap();
+    /// assert_eq!(dt.to_string(), "2022-01-01T12:13:14Z");
+    /// ```
+    pub fn parse_bytes_with_config(bytes: &[u8], _config: TimeConfig) -> Result<Self, ParseError> {
```
[the remark this answers is not in the document]
Comment (adriangb) [reply src/datetime.rs:291]: We'd be breaking anything that uses this! It wasn't too much harder. Also, Rust has no default arguments...

File: src/time.rs [hunk 4]
```diff
@@ -488,3 +523,22 @@ impl PureTime {
...
 }
+
+#[derive(Debug, Clone)]
+pub enum SecondsPrecisionOverflowBehavior {
+    Truncate,
+    Error,
+}
+
+#[derive(Debug, Clone)]
+pub struct TimeConfig {
```
[the remark this answers is not in the document]
Comment (adriangb) [reply src/time.rs:534]: Okay happy to do that. I'm just concerned that then if we end up putting `date` stuff in here we'll be passing `date` options to `Time` which has no concept of dates.

File: src/time.rs (hunk 2 above)
Comment (adriangb) [reply src/time.rs]: done

File: src/time.rs (hunk 1 above)
Comment (adriangb) [reply src/time.rs:532]: done

File: src/time.rs [hunk 5]
```diff
@@ -450,16 +481,23 @@ impl PureTime {
                     loop {
                         match bytes.get(offset + length + i) {
                             Some(c) if c.is_ascii_digit() => {
-                                microsecond *= 10;
-                                microsecond += (c - b'0') as u32;
+                                if i < 6 {
+                                    microsecond *= 10;
+                                    microsecond += (c - b'0') as u32;
+                                }
```
Comment (adriangb) [inline src/time.rs:487]:
```suggestion
                               // If we've passed `i=6` then we are "truncating" the extra precision
                               // The easiest way to do this is to simply no-op and continue the loop
                                if i < 6 {
                                    microsecond *= 10;
                                    microsecond += (c - b'0') as u32;
                                }
```

**Comment by adriangb** [comment]
please review

**Review by davidhewitt — APPROVED** [review]

**Pull request merged**